flux-tests/tests/pos/surface/range.rs

```diff
@@ -0,0 +1,35 @@
+#![feature(register_tool)]
+#![register_tool(flux)]
+
+#[flux::refined_by(lo: int, hi: int)]
+pub struct Rng {
+    #[flux::field(i32[@lo])]
+    _lo: i32,
+    #[flux::field(i32[@hi])]
+    hi: i32,
+    #[flux::field(i32{v:lo <= v && v <= hi})]
+    cur: i32,
+}
+
+impl Rng {
+    #[flux::sig(fn(lo:i32, hi:i32{lo <= hi}) -> Rng[lo, hi])]
+    pub fn new(lo: i32, hi: i32) -> Self {
+        Self { _lo: lo, hi, cur: lo }
+    }
+}
+
+impl Iterator for Rng {
+    type Item = i32;
+
+    #[flux::sig(fn(self: &mut Rng[@lo, @hi]) -> Option<i32{v:lo <= v && v < hi}>)]
+    fn next(&mut self) -> Option<i32> {
```

Comment:
I think we had this conversation, but what else is needed to give this a stronger type that says the return value is exactly `Some(cur)` as long as the range hasn't been exhausted?

Not for this PR.

Reply:
Well, we could technically use the `cur`-indexed iterator and then just say that its `cur`? 

```
#[flux::sig(fn(self: &mut RngIter[@lo, @cur, @hi]) -> Option<i32[cur]>]
```

Reply:
We would also need to make the reference strong and update `curr` at the output, but you should only update it when returning `Some`.

This also makes me think to what extent is sound to make it a strong reference here but a mutable reference in the trait definition.